nncf/experimental/onnx/graph/model_transformer.py

```diff
@@ -148,93 +149,117 @@ def select_model_inputs_outputs(model, outputs=None):
+        zero_point_tensor_name = ONNXModelTransformer.ZERO_POINT_NAME_PREFIX + q_target_edge_name
 
         quantizer = onnx.helper.make_node(
-            'QuantizeLinear',
-            inputs=[target_edge_name, scale_tensor_name, zero_point_tensor_name],
-            outputs=['q_output_' + target_edge_name],
             name=quantizer_name,
+            op_type='QuantizeLinear',
+            inputs=[input_target_edge, scale_tensor_name, zero_point_tensor_name],
+            outputs=['q_output_' + q_target_edge_name],
             axis=axis
         )
-        # If several nodes on one edge
-        dequantizer_outputs = ['dq_output_' + st for st in target_edge_names]
+
         dequantizer = onnx.helper.make_node(
-            'DequantizeLinear',
-            inputs=['q_output_' + target_edge_name, scale_tensor_name, zero_point_tensor_name],
-            outputs=dequantizer_outputs,
             name=dequantizer_name,
+            op_type='DequantizeLinear',
+            inputs=['q_output_' + q_target_edge_name, scale_tensor_name, zero_point_tensor_name],
+            outputs=['dq_output_' + q_target_edge_name],
             axis=axis,
-
         )
 
-        # TODO (kshpv): need to carefully look through the logic of nodes searching.
-        #  The model with the possible issues is inception_v3.
-        # If several nodes on one edge
-        for target_edge_name in target_edge_names:
-            input_nodes = onnx_graph.get_nodes_by_input(target_edge_name)
-            if not input_nodes:
-                raise RuntimeError(
-                    f'Can not add the quantizer to the {target_edge_name} edge. This edge does not have end node.')
+        return quantizer, dequantizer
+
+    def _get_scale_zero_point_tensors(self, transformation, quantizer, dequantizer):
+        scale = transformation.quantizer_parameters.scale
+        zero_point = transformation.quantizer_parameters.zero_point
+        mode = transformation.quantizer_parameters.mode
+
+        per_channel = isinstance(scale, list)
+
+        zero_point = [zero_point] if not isinstance(zero_point, list) else zero_point
+
+        scale = [scale] if not isinstance(scale, list) else scale
+        tensor_type = onnx.TensorProto.UINT8 if mode == QuantizationMode.ASYMMETRIC else onnx.TensorProto.INT8
+        dims = [len(scale)] if per_channel else []
+        assert quantizer.input[1] == dequantizer.input[1] and quantizer.input[2] == dequantizer.input[2]
+        scale_tensor_name = quantizer.input[1]
+        zero_point_tensor_name = quantizer.input[2]
+
+        onnx_scale = onnx.helper.make_tensor(scale_tensor_name, onnx.TensorProto.FLOAT, dims, scale)
+        onnx_zero_point = onnx.helper.make_tensor(zero_point_tensor_name, tensor_type, dims, zero_point)
+        return onnx_scale, onnx_zero_point
+
+    def _insert_quantizer_dequantizer(self, transformation: ONNXQuantizerInsertionCommand):
+        onnx_graph = ONNXGraph(self.transformed_model)
```

Comment:
Can we create `ONNXGraph` once, outside of the loop of transformations?

Reply:
Maybe we could. It could be done in separate PR

Reply:
I checked the `ONNXGraph`'s `__init__` method and it seems that the initialization does quite a lot of work - inferring shapes, populating the tensor names etc, doing at least O(N) work where N is the number of nodes in the graph. By building the ONNXGraph for every quantizer insertion you turn the work of inserting all quantizers into the graph from O(N) to O(N^2), which is unacceptable. Optimize your code in this PR utilizing @KodiaqQ's suggestion.

Reply:
So because of that it has TODO: optimize. 
The reason why it couldn't easy to change beacuse every insertion of quantizer/dequantizer changes the model. For safe reason the graph recreates every time.

Reply:
What safety are you talking about exactly? Does the model change during an insertion of a quantizer make it impossible to insert other quantizers at the points determined before, do target edge addresses in ONNX graph change every time you insert a quantizer anywhere in the network? Please post a screenshot or a model file of a minimal example where adding one quantizer to a given place makes the address for placing the second quantizer invalid.

Reply:
The ONNX format demands that the nodes are topologically sorted. So when I put the new node I use the insert_index, which is obtained from ONNXGraph. If I don't update the model's indexes it can cause the error

```
        self.transformed_model.graph.initializer.extend([onnx_scale])
        self.transformed_model.graph.initializer.extend([onnx_zero_point])
        insert_index = onnx_graph.get_node_index(input_nodes[0].name)
        self.transformed_model.graph.node.insert(insert_index, quantizer)
        self.transformed_model.graph.node.insert(insert_index + 1, dequantizer)
```

Reply:
After offline discussion we concluded to make a separate PR optimizing ONNXGraph creation. 
Which what I was going to do.